### PR TITLE
fix(hash): allow characters after `#` in base e.g. `#!`

### LINKED ISSUE #923: Base URL with a hashbang rewrites the URL breaking the routing
### Version
4.0.6

### Reproduction link
[https://codesandbox.io/s/keen-cannon-tmkfz](https://codesandbox.io/s/keen-cannon-tmkfz)

### Steps to reproduce
If you click on a link it works and renders the correct view and URL updates accordingly but if you can't refresh the URL as the createCurrentLocation method looks for a hash and places a forward slash "/" right after it which brakes the hashbang URL structure.

### What is expected?
If you refresh a route that has been rendered using routing the page should be able to be re-rendered based on the URL

### What is actually happening?
The URL is being updated before it can be matched and so no route is found for the URL

---
While not a recommended best practice hashbang urls are still valid and should be rather trivial to support.  I specifically have an app that I am re-writing and in order to not break URLs need to support the hashbang structure

### PR DESCRIPTION
**Pull request by GGwujun**
Fix #923

### CODE REVIEW

**Comment by edison1105** [comment]
duplicate of this PR: #924

**Comment by GGwujun** [comment]
See also [docs](https://developers.google.com/search/docs/ajax-crawling/docs/getting-started).

Fix #923

**Review by posva — CHANGES_REQUESTED** [review]
Thanks! Could you add a unit test to handle these bases

- `#!`
- `#other`

File: src/history/html5.ts [hunk 1]
```diff
@@ -39,11 +39,15 @@ function createCurrentLocation(
...
 ): HistoryLocation {
   const { pathname, search, hash } = location
-  // allows hash based url
+  // allows hash based or hashbang based url or others
   const hashPos = base.indexOf('#')
   if (hashPos > -1) {
-    // prepend the starting slash to hash so the url starts with /#
-    let pathFromHash = hash.slice(1)
+    let slicePos =
+      hash.indexOf(base.slice(hashPos).toLowerCase()) > -1
```
Comment (posva) [inline src/history/html5.ts:46]:
```suggestion
      hash.includes(base.slice(hashPos))
```

File: src/history/html5.ts [hunk 2]
```diff
@@ -39,11 +39,15 @@ function createCurrentLocation(
   location: Location
 ): HistoryLocation {
   const { pathname, search, hash } = location
-  // allows hash based url
+  // allows hash based or hashbang based url or others
   const hashPos = base.indexOf('#')
   if (hashPos > -1) {
-    // prepend the starting slash to hash so the url starts with /#
```
Comment (posva) [inline src/history/html5.ts:45]: can you add back the comment at line 50

**Review by posva — CHANGES_REQUESTED** [review]
Thanks!

**Comment by posva** [comment]
I added tests that were failing before. They seem to be failing when ran with other tests but work when ran alone. It will need more investigation

File: src/history/html5.ts [hunk 3]
```diff
@@ -39,11 +39,15 @@ function createCurrentLocation(
...
-    let pathFromHash = hash.slice(1)
+    let slicePos =
+      hash.indexOf(base.slice(hashPos).toLowerCase()) > -1
+        ? base.slice(hashPos).length
+        : 1
+    let pathFromHash = hash.slice(slicePos)
+
```
Comment (posva) [inline src/history/html5.ts:50]:
```suggestion
// prepend the starting slash to hash so the url starts with /#
```

File: src/history/html5.ts (hunk 1 above)
Comment (GGwujun) [reply src/history/html5.ts]: good,thks

**Review by posva — APPROVED** [review]

**Pull request merged**